File: medium/3629. Minimum Jumps to Reach End via Prime Teleportation/solution.rs
use std::collections::{VecDeque, HashMap};

impl Solution {
    pub fn min_jumps(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut maxv = 0;
        for &x in &nums { if x > maxv { maxv = x; } }
        let maxv = maxv as usize;

        let mut spf = vec![0usize; maxv + 1];
        for i in 2..=maxv {
            if spf[i] == 0 {
                spf[i] = i;
                if i * i <= maxv {
                    let mut j = i * i;
                    while j <= maxv {
                        if spf[j] == 0 { spf[j] = i; }
                        j += i;
                    }
                }
            }
        }

        fn factors(mut x: usize, spf: &Vec<usize>) -> Vec<usize> {
            let mut r = Vec::new();
            while x > 1 {
                let p = spf[x];
                r.push(p);
                while x % p == 0 { x /= p; }
            }
            r
        }

        let mut mp: HashMap<usize, Vec<usize>> = HashMap::new();
        let mut facs = Vec::with_capacity(n);
        for i in 0..n {
            let fs = factors(nums[i] as usize, &spf);
            for &p in &fs {
                mp.entry(p).or_default().push(i);
            }
            facs.push(fs);
        }

        let mut vis = vec![false; n];
        let mut q = VecDeque::new();
        vis[0] = true;
        q.push_back((0usize, 0i32));

        while let Some((i, d)) = q.pop_front() {
            if i == n - 1 { return d; }

            if i + 1 < n && !vis[i + 1] {
                vis[i + 1] = true;
                q.push_back((i + 1, d + 1));
            }
            if i > 0 && !vis[i - 1] {
                vis[i - 1] = true;
                q.push_back((i - 1, d + 1));
            }

            let v = nums[i] as usize;

            if spf[v] == v {
                for &p in &facs[i] {
                    if let Some(list) = mp.get_mut(&p) {
                        for &j in list.iter() {
                            if !vis[j] {
                                vis[j] = true;
                                q.push_back((j, d + 1));
                            }
                        }
                        list.clear();
                    }
                }
            }
        }

        -1
    }
}

File: medium/3629. Minimum Jumps to Reach End via Prime Teleportation/solution.cpp
/*
Time: O(n * sqrt(max(nums)))
Space: O(n)

Idea: BFS + prime factor graph
*/

class Solution {
public:
    vector<int> getFactors(int x, vector<int>& spf) {
        vector<int> res;

        while (x > 1) {
            int p = spf[x];
            res.push_back(p);

            while (x % p == 0) {
                x /= p;
            }
        }

        return res;
    }

    int minJumps(vector<int>& nums) {
        int n = nums.size();

        int mx = *max_element(nums.begin(), nums.end());

        // smallest prime factor
        vector<int> spf(mx + 1);

        for (int i = 2; i <= mx; i++) {
            if (spf[i] == 0) {
                spf[i] = i;

                if (1LL * i * i <= mx) {
                    for (long long j = 1LL * i * i; j <= mx; j += i) {
                        if (spf[j] == 0) {
                            spf[j] = i;
                        }
                    }
                }
            }
        }

        unordered_map<int, vector<int>> mp;
        vector<vector<int>> facs(n);

        // store factors for each index
        for (int i = 0; i < n; i++) {
            facs[i] = getFactors(nums[i], spf);

            for (int p : facs[i]) {
                mp[p].push_back(i);
            }
        }

        vector<bool> vis(n, false);

        queue<pair<int,int>> q;
        q.push({0, 0});
        vis[0] = true;

        while (!q.empty()) {
            auto [i, d] = q.front();
            q.pop();

            if (i == n - 1) {
                return d;
            }

            // move right
            if (i + 1 < n && !vis[i + 1]) {
                vis[i + 1] = true;
                q.push({i + 1, d + 1});
            }

            // move left
            if (i > 0 && !vis[i - 1]) {
                vis[i - 1] = true;
                q.push({i - 1, d + 1});
            }

            int v = nums[i];

            // jump using prime factors
            if (spf[v] == v) {
                for (int p : facs[i]) {
                    auto &list = mp[p];

                    for (int j : list) {
                        if (!vis[j]) {
                            vis[j] = true;
                            q.push({j, d + 1});
                        }
                    }

                    list.clear();
                }
            }
        }

        return -1;
    }
};